### PR TITLE
training-on-multiple-scenarios, InfoGAIL

### PR DESCRIPTION
**Pull request by YuechengLiu**
- training on multiple scenarios (ngsim-i80, ngsim-us101...)
- InfoGAIL support
- smarts API change
- support to train on time-sliced scenarios (cut-in cases, etc.)

### CODE REVIEW

File: ILSwiss/exp_specs/evaluate/visualize_gail_smarts.yaml [hunk 1]
```diff
@@ -10,23 +10,27 @@ variables:
...
 # -----------------------------------------------------------------------------
 constants:
-  expert_name: 'smarts_i80'
+  expert_name: 'smarts_i80_radius_stack_1'
   expert_idx: 0
   traj_num: -1
   scale_env_with_demo_stats: true
 
-  log_path: "logs/gail-smarts-ngsim--scale--terminal--gp-4.0--rs-2.0--trajnum--1/gail_smarts_ngsim--scale--terminal--gp-4.0--rs-2.0--trajnum--1_2022_03_18_09_27_17_0000--s-0"
+  log_path: "/home/liuyuecheng/i80_fixbug_result/buf2M_dis64x2_policy_512x3_r2_alpha1"
```
Comment (zbzhu99) [inline ILSwiss/exp_specs/evaluate/visualize_gail_smarts.yaml]: remove absolute path?

File: ILSwiss/rlkit/core/eval_util.py [hunk 2]
```diff
@@ -10,9 +10,216 @@
...
+class ScenarioWiseStats:
+    _attr_mapping = {
+        "success_rate": "win_count",
+        "collision_rate": "collision_rate",
+        "collision_time": "collision_time",
+        "dist_to_hist_cur_pos_sum": "dist_to_hist_cur_pos_sum",
+        "dist_to_hist_cur_pos_mean": "dist_to_hist_cur_pos_mean",
+        "dist_to_hist_final_pos": "dist_to_hist_final_pos",
+        "lane_change": "lane_change",
+    }
```
Comment (zbzhu99) [inline ILSwiss/rlkit/core/eval_util.py]: Why we need a mapping just to change `success_rate` to `win_count`?

File: ILSwiss/rlkit/data_management/path_builder.py [hunk 3]
```diff
@@ -2,8 +2,10 @@
 
 
 class PathBuilder:
-    def __init__(self, agent_ids):
+    def __init__(self, agent_ids, scenario_name=None, traffic_name=None):
```
Comment (zbzhu99) [inline ILSwiss/rlkit/data_management/path_builder.py:5]: Why pass `scenario_name` and `traffic_name` to PathBuilder?

File: ILSwiss/rlkit/samplers/__init__.py [hunk 4]
```diff
@@ -131,29 +142,27 @@ def obtain_samples(self, num_steps=None, pred_obs=False):
...
-                continue
-            _actions = self.policy_n[policy_id].get_actions(
-                np.stack(_observations, axis=0)
-            )
-            for idx, action in zip(_idxes, _actions):
-                action_n[idx][agent_id] = action
-        return action_n
+        actions_n = [{} for _ in range(len(observations_n))]
+        policy_obs_list_n = defaultdict(list)  # to recover the corresponding actions
+        policy_to_env_agent_id = defaultdict(list)
```
Comment (zbzhu99) [inline ILSwiss/rlkit/samplers/__init__.py:151]: Make the mapping from p_id to a_id an attribute of PathSampler to avoid recomputing?

File: ILSwiss/rlkit/torch/algorithms/info_adv_irl/info_adv_irl.py [hunk 5]
```diff
@@ -0,0 +1,645 @@
...
+        disc_lr=1e-3,
+        disc_focal_loss_gamma=0.0,
+        disc_momentum=0.9,
+        disc_optimizer_class=optim.Adam,
+        post_r_coef=0.1,
+        latent_distribution=None,
+        use_grad_pen=True,
+        grad_pen_weight=10,
+        rew_clip_min=None,
+        rew_clip_max=None,
```
Comment (zbzhu99) [inline ILSwiss/rlkit/torch/algorithms/info_adv_irl/info_adv_irl.py]: Duplicate arguments with AdvIRL can be merged into kwargs.

File: ILSwiss/rlkit/torch/algorithms/info_adv_irl/info_adv_irl.py [hunk 6]
```diff
@@ -0,0 +1,645 @@
...
+        batch = self.get_batch(self.posterior_optim_batch_size, agent_id, False)
+        # print(f"batch.keys: {batch.keys()}")
+        self.posterior_trainer_n[policy_id].train_step(batch)
+
+    @property
+    def networks_n(self):
+        return {
+            p_id: [self.discriminator_n[p_id]] + self.policy_trainer_n[p_id].networks
+            for p_id in self.policy_ids
+        }
```
Comment (zbzhu99) [inline ILSwiss/rlkit/torch/algorithms/info_adv_irl/info_adv_irl.py]: Do we need to save posterior networks here?

File: ILSwiss/rlkit/torch/algorithms/info_adv_irl/info_adv_irl.py [hunk 7]
```diff
@@ -0,0 +1,645 @@
...
+            disc_momentum=disc_momentum,
+            disc_optimizer_class=disc_optimizer_class,
+            use_grad_pen=use_grad_pen,
+            grad_pen_weight=grad_pen_weight,
+            rew_clip_min=rew_clip_min,
+            rew_clip_max=rew_clip_max,
+            **kwargs,
+        )
+
+        self.eval_sampler._init_latent_distribution(latent_distribution)
```
Comment (zbzhu99) [inline ILSwiss/rlkit/torch/algorithms/info_adv_irl/info_adv_irl.py]: Methods staring with _ is considered to be protected. Maybe change to `init_latent_distribution()` or wrap `_init_latent_distribution` inside a exposed method like `init()`.

File: ILSwiss/rlkit/torch/algorithms/info_adv_irl/info_adv_irl.py [hunk 8]
```diff
@@ -0,0 +1,645 @@
...
+        if self.eval_statistics is None:
+            self.eval_statistics = OrderedDict()
+        for p_id in self.policy_ids:
+            posterior_statistics = self.posterior_trainer_n[p_id].get_eval_statistics()
+            for name, data in posterior_statistics.items():
+                self.eval_statistics.update({f"{p_id} {name}": data})
+
+        # AdvIRL evaluate
+        if self.eval_statistics is None:
+            self.eval_statistics = OrderedDict()
```
Comment (zbzhu99) [inline ILSwiss/rlkit/torch/algorithms/info_adv_irl/info_adv_irl.py]: This can be safely removed.

File: ILSwiss/rlkit/torch/algorithms/sac/conditional_sac_alpha.py [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+import numpy as np
+import torch
+import torch.optim as optim
+
+import rlkit.torch.utils.pytorch_util as ptu
+from rlkit.core.trainer import Trainer
+from rlkit.core.eval_util import create_stats_ordered_dict
+
+
+class ConditionalSoftActorCritic(Trainer):
```
Comment (zbzhu99) [inline ILSwiss/rlkit/torch/algorithms/sac/conditional_sac_alpha.py]: Can this class be inherited from SoftActorCritic?

File: ILSwiss/rlkit/torch/algorithms/sac/sac.py [hunk 10]
```diff
@@ -72,7 +74,10 @@ def train_step(self, batch):
         # policy_params = itertools.chain(self.policy.parameters())
 
         rewards = self.reward_scale * batch["rewards"]
-        terminals = batch["terminals"]
+        if self.ignore_terminal:
+            terminals = 0.0
+        else:
+            terminals = batch["terminals"]
```
Comment (zbzhu99) [inline ILSwiss/rlkit/torch/algorithms/sac/sac.py]: Why we need to ignore terminal in SAC training?

File: ILSwiss/rlkit/torch/algorithms/sac/sac_alpha.py [hunk 11]
```diff
@@ -79,7 +81,10 @@ def train_step(self, batch):
         # policy_params = itertools.chain(self.policy.parameters())
 
         rewards = self.reward_scale * batch["rewards"]
-        terminals = batch["terminals"]
+        if self.ignore_terminal:
+            terminals = 0.0
+        else:
+            terminals = batch["terminals"]
```
Comment (zbzhu99) [inline ILSwiss/rlkit/torch/algorithms/sac/sac_alpha.py]: Same as in `sac.py`

File: ILSwiss/rlkit/torch/common/networks.py [hunk 12]
```diff
@@ -101,6 +97,184 @@ def forward(self, input, return_preactivations=False):
...
+        assert self.layer_norm is False
+        assert self.batch_norm is False
+        assert self.batch_norm_before_output_activation is False
+        h = input
+        for i, fc in enumerate(self.fcs):
+            h = fc(h)
+            h = self.hidden_activation(h)
+        preactivation = self.last_fc(h)
+        output = self.output_activation(preactivation)
+        return output
```
Comment (zbzhu99) [inline ILSwiss/rlkit/torch/common/networks.py:111]: Just curious to ask, how much acceleration does jit roughly provide?
and what is the minimal version of PyTorch required to use jit?

File: smarts-imitation/smarts_imitation/envs/core.py [hunk 13]
```diff
@@ -247,23 +295,149 @@ def _init_scenario(self):
...
+        query = """SELECT T.position_x, T.position_y, T.sim_time
+                   FROM Trajectory AS T
+                   WHERE T.vehicle_id = ? AND T.sim_time > ? AND T.sim_time <= ? 
+                   ORDER BY T.sim_time DESC"""
+        rows = self.scenario._traffic_history._query_list(
+            query, (vehicle_id, start_time, end_time)
+        )
+        return (VehiclePosition(*row) for row in rows)
+
+    def _get_vehicle_current_history_position(self, vehicle_id):
```
Comment (zbzhu99) [inline smarts-imitation/smarts_imitation/envs/core.py:343]: Please add comments to briefly explain the meaning of `current_history_position`.

File: smarts-imitation/smarts_imitation/envs/core.py [hunk 14]
```diff
@@ -247,23 +295,149 @@ def _init_scenario(self):
...
+            self.time_slice = False
+            self.vehicle_missions = (
+                self.scenario.discover_missions_of_traffic_histories()
+            )
+            self.vehicle_ids = list(self.vehicle_missions.keys())
+            self.vehicle_start_times = {
+                v_id: self.vehicle_missions[v_id].start_time
+                for v_id in self.vehicle_ids
+            }
+        elif self.vehicles[0].start_time is None:
```
Comment (zbzhu99) [inline smarts-imitation/smarts_imitation/envs/core.py:447]: I wonder in what cases self.vehicles[0].start_time is None or not.

File: smarts-imitation/smarts_imitation/utils/common.py [hunk 15]
```diff
@@ -226,11 +226,13 @@ def cal_neighbor_with_radius_ego_coordinate(env_obs: Observation, **kwargs):
...
             if v[0] is None:
-                dist = 15.0
+                # dist = 15.0
+                dist = -5.0
                 heading_vec = [0.0, 0.0]
                 speed = -1.0
             else:
-                dist = min(v[1], 15.0)
+                # dist = min(v[1], 15.0)
+                dist = v[1]
```
Comment (zbzhu99) [inline smarts-imitation/smarts_imitation/utils/common.py:235]: What is the change here for?

**Review by zbzhu99 — CHANGES_REQUESTED** [review]

File: ILSwiss/exp_specs/evaluate/visualize_gail_smarts.yaml (hunk 1 above)
Comment (YuechengLiu) [reply ILSwiss/exp_specs/evaluate/visualize_gail_smarts.yaml]: I will remove them

File: ILSwiss/rlkit/data_management/path_builder.py (hunk 3 above)
Comment (YuechengLiu) [reply ILSwiss/rlkit/data_management/path_builder.py:5]: It is used to build scenario-wise or traffic-wise stats in `eval_util`
https://github.com/YuechengLiu/NGSIM_Imitation/blob/6c1e8807aeaca9738a8548b4bb54d1603968cf80/ILSwiss/rlkit/core/eval_util.py#L59

File: ILSwiss/rlkit/torch/common/networks.py (hunk 12 above)
Comment (YuechengLiu) [reply ILSwiss/rlkit/torch/common/networks.py:111]: `torch.jit` is not used for training. It is used only for saving a script model for C++ inference. See `run_scripts/to_cpp_model.py` for details.

File: smarts-imitation/smarts_imitation/utils/common.py (hunk 15 above)
Comment (YuechengLiu) [reply smarts-imitation/smarts_imitation/utils/common.py:235]: It performs better in my experiments (remove the max distance clipping and zero-padding the vehicle-no-found cases)

File: ILSwiss/rlkit/torch/algorithms/info_adv_irl/info_adv_irl.py (hunk 6 above)
Comment (YuechengLiu) [reply ILSwiss/rlkit/torch/algorithms/info_adv_irl/info_adv_irl.py]: Yes, you are right. We need also save the posterior networks.

Comment (YuechengLiu) [reply ILSwiss/rlkit/torch/algorithms/info_adv_irl/info_adv_irl.py]: I notice that another `networks_n` is defined and I forget to remove the old one.

File: ILSwiss/rlkit/core/eval_util.py (hunk 2 above)
Comment (YuechengLiu) [reply ILSwiss/rlkit/core/eval_util.py]: set `ScenarioWiseStats._attrs` as a list.

File: ILSwiss/rlkit/torch/algorithms/info_adv_irl/info_adv_irl.py (hunk 5 above)
Comment (YuechengLiu) [reply ILSwiss/rlkit/torch/algorithms/info_adv_irl/info_adv_irl.py]: fixed

File: ILSwiss/rlkit/torch/algorithms/info_adv_irl/info_adv_irl.py (hunk 7 above)
Comment (YuechengLiu) [reply ILSwiss/rlkit/torch/algorithms/info_adv_irl/info_adv_irl.py]: fixed

File: ILSwiss/rlkit/torch/algorithms/sac/conditional_sac_alpha.py (hunk 9 above)
Comment (YuechengLiu) [reply ILSwiss/rlkit/torch/algorithms/sac/conditional_sac_alpha.py]: fixed

File: ILSwiss/rlkit/torch/algorithms/sac/sac_alpha.py (hunk 11 above)
Comment (YuechengLiu) [reply ILSwiss/rlkit/torch/algorithms/sac/sac_alpha.py]: I tried to use this option to alleviate the terminal effect during training. Now I have removed the option.

File: smarts-imitation/smarts_imitation/envs/core.py (hunk 13 above)
Comment (YuechengLiu) [reply smarts-imitation/smarts_imitation/envs/core.py:343]: added.

File: smarts-imitation/smarts_imitation/envs/core.py (hunk 14 above)
Comment (YuechengLiu) [reply smarts-imitation/smarts_imitation/envs/core.py:447]: I have added more comments.

Comment (zbzhu99) [reply smarts-imitation/smarts_imitation/envs/core.py:447]: So why self.vehicles[0].start_time can be None? It seems to be the time that the vehicle first appears.

File: ILSwiss/rlkit/torch/algorithms/sac/sac_alpha.py [hunk 16]
```diff
@@ -116,7 +116,7 @@ def train_step(self, batch):
         )  # do not need grad || it's the shared part of two calculation
         min_target_value = torch.min(target_qf1_values, target_qf2_values)
         q_target = rewards + (1.0 - terminals) * self.discount * (
-            min_target_value - self.alpha * next_log_pi
+            min_target_value - self.alpha * self.reward_scale * next_log_pi
```
Comment (zbzhu99) [inline ILSwiss/rlkit/torch/algorithms/sac/sac_alpha.py]: Why multiply self.reward_scale here? I am not sure if this is reasonable.

Comment (YuechengLiu) [reply ILSwiss/rlkit/torch/algorithms/sac/sac_alpha.py]: The `Discriminator reward` is scaled by `self.reward_scale`, and I think that the `max entropy reward` should also keep the same scale.

File: smarts-imitation/smarts_imitation/envs/core.py (hunk 14 above)
Comment (YuechengLiu) [reply smarts-imitation/smarts_imitation/envs/core.py:447]: In the case that self.vehicles[0].start_time is `None`, user does not specify the `start_time` and `end_time` of each vehicle. So the default start time is used (first seen time) to initialize the vehicles. 

In fact, the `self.vehicles[0].start_time` and `self.vehicles[0].end_time` is only used for the `cut-in cases training` now. (I just filter out some cut-in cases from the traffic histories and save the start_time and end_time and the corresponding vehicle_id and replay them during trianing.)

File: ILSwiss/rlkit/samplers/__init__.py (hunk 4 above)
Comment (YuechengLiu) [reply ILSwiss/rlkit/samplers/__init__.py:151]: According to my understanding, it can not save too much time. (Maybe I'm wrong and you can try to improve it if possible)

Comment (zbzhu99) [reply ILSwiss/rlkit/samplers/__init__.py:151]: Sure, this is not a time-related issue and I just think making the computing as a preprocessing step is more natural. Keep it like this is okay.

File: ILSwiss/rlkit/torch/algorithms/sac/sac_alpha.py (hunk 16 above)
Comment (zbzhu99) [reply ILSwiss/rlkit/torch/algorithms/sac/sac_alpha.py]: @Ericonaldo What do you think

File: smarts-imitation/smarts_imitation/envs/core.py (hunk 14 above)
Comment (zbzhu99) [reply smarts-imitation/smarts_imitation/envs/core.py:447]: Ok, thanks for your explanation.

File: ILSwiss/rlkit/torch/algorithms/sac/sac_alpha.py (hunk 16 above)
Comment (Ericonaldo) [reply ILSwiss/rlkit/torch/algorithms/sac/sac_alpha.py]: Although I think in practice it will make less difference, here I think we should not scale the entropy. Because the reward scale helps construct the reward function only, and the other should be the same as the regular RL algorithm.

Comment (YuechengLiu) [reply ILSwiss/rlkit/torch/algorithms/sac/sac_alpha.py]: In sac,  the entropy term is also one part of the value function, and the optimal policy is only depending on the ratio between `original reward term` and `entropy reward term` (regardless of their magnitude). If we just only scale the original reward term, the optimal policy is depending on both `self.alpha` and `self.reward_scale`, and this is not convenient to control the exploration/exploitation trade-off (policy entropy regularization). Under my design, the reward scale is only controlled by `self.reward_scale` and exploration/exploitation trade-off is only controlled by `self.alpha`. 

But it is ok to revert the code to original version. It is just a minor problem.

**Review by CyberSY — COMMENTED** [review]
No big issues as I reviewed
